executor/aggfuncs/func_count.go

```diff
@@ -0,0 +1,110 @@
+package aggfuncs
+
+import (
+	"github.com/juju/errors"
+
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+type baseCount struct {
+	baseAggFunc
+}
+
+type partialResult4Count = int64
+
+func (e *baseCount) AllocPartialResult() PartialResult {
+	return PartialResult(new(partialResult4Count))
+}
+
+func (e *baseCount) ResetPartialResult(pr PartialResult) {
+	p := (*partialResult4Count)(pr)
+	*p = 0
+}
+
+func (e *baseCount) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	p := (*partialResult4Count)(pr)
+	chk.AppendInt64(e.ordinal, *p)
+	return nil
+}
+
+type countOriginal struct {
+	baseCount
+}
+
+func (e *countOriginal) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) error {
+	p := (*partialResult4Count)(pr)
+
+	for _, row := range rowsInGroup {
+		isNull, err := e.args[0].IsNull(sctx, row)
```

Comment:
We don't need to implement a `IsNull` api on the `Expression` interface, and we'd better not to do this:
- this API is currently only used in `COUNT` function
- in the future, we are planning to change the aggregate argument from `args []expression.Expression` to `args []*expression.Column`

There are two ways to solve the `NULL` input issue:
1. use a single `countOriginal` to handle all the input types, and:
```go
switch e.args[i].GetType().EvalType() {
case types.ETInt:
    val, isNull, err := e.args[0].EvalInt(ctx, row)
    if err != nil {
        return errors.Trace(err)
    }
    if isNull {
        continue
    }
case types.ETReal:
        ...
}
```

 2. implement a `countOriginal4XX` for each input type:
 - `countOriginal4Decimal`
 - `countOriginal4Int64`
 - ...

I suggest the second way: It reduces a lot of CPU branch predications, which are introduced by the `switch` statement, utilizes the CPU pipelines thus is more inefficient than the first way.